### ISSUE TITLE
Query source has already been associated with an expression | LTS and Workaround?

### ISSUE THREAD

**Issue by spottedmahn** [dotnet/efcore#15700]
Appears to be the same issue as #11900...

```
Exception message: 'Query source (from blah <generated>_1 in [g]) has already been associated with an expression.'
Stack trace:
   at Remotion.Linq.Clauses.QuerySourceMapping.AddMapping(IQuerySource querySource, Expression expression)
   at Remotion.Linq.Clauses.MainFromClause.Clone(CloneContext cloneContext)
   at Remotion.Linq.QueryModel.Clone(QuerySourceMapping querySourceMapping)
   at Remotion.Linq.QueryModel.CloningExpressionVisitor.VisitSubQuery(SubQueryExpression expression)
   at Remotion.Linq.Clauses.Expressions.SubQueryExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitAndConvert[T](ReadOnlyCollection`1 nodes, String callerName)
   at Remotion.Linq.Parsing.RelinqExpressionVisitor.VisitNew(NewExpression expression)
   at System.Linq.Expressions.NewExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Remotion.Linq.Clauses.SelectClause.TransformExpressions(Func`2 transformation)
   at Remotion.Linq.QueryModel.TransformExpressions(Func`2 transformation)
   at Remotion.Linq.QueryModel.Clone(QuerySourceMapping querySourceMapping)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryOptimizer.FlattenSubQuery(SubQueryExpression subQueryExpression, IFromClause fromClause, QueryModel queryModel, Int32 destinationIndex)
   at Remotion.Linq.Transformations.SubQueryFromClauseFlattener.VisitMainFromClause(MainFromClause fromClause, QueryModel queryModel)
   at Remotion.Linq.Clauses.MainFromClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryOptimizer.Optimize(QueryCompilationContext queryCompilationContext, QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.OptimizeQueryModel(QueryModel queryModel, Boolean asyncQuery)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.OptimizeQueryModel(QueryModel queryModel, Boolean asyncQuery)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateAsyncQueryExecutor[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.System.Collections.Generic.IAsyncEnumerable<TResult>.GetEnumerator()
   at System.Linq.AsyncEnumerable.Aggregate_[TSource,TAccumulate,TResult](IAsyncEnumerable`1 source, TAccumulate seed, Func`3 accumulator, Func`2 resultSelector, CancellationToken cancellationToken)
```

### Workaround
I'm having a hard time understanding the workaround in #11900.

>you can try rewriting the query so that collection navigation projection is the last operation - if you have OrderBy, Where, etc afterwards

[source](https://github.com/aspnet/EntityFrameworkCore/issues/11900#issuecomment-444991415)

I don't think I'm doing any navigation projection in the following, right?

```csharp
//'Query source (from Blah <generated>_1 in [g]) has already been associated with an expression.'
var query = await db.Blah
    .GroupBy(x => x.VersionId)
    .Select(g => new { g.Key, Count = g.Count() })
    .OrderByDescending(z => z.Count)
    .Take(50)
    .Select(m => m.Key)
    .ToListAsync();

//workaround
var query = await db.Blah
	.GroupBy(x => x.VersionId)
	.Select(g => new { g.Key, Count = g.Count() })
	.ToListAsync();

var workaround = query.OrderByDescending(z => z.Count)
	.Take(50)
	.Select(m => m.Key)
	.ToList();
```

### Further technical details
EF Core version: `<PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.10"/>`
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Win10
IDE: Visual Studio 2019 16.1.0 Preview 3.0

### Questions
- I see this is targeted to be fixed in 3.0, correct?  What about 2.1?  If 2.1 is LTS then shouldn't it also be patched there?
- What is the proper workaround here?

**Comment by ajcvickers** [comment]
@spottedmahn The workaround you show above is what we would recommend as a workaround for this issue, so we're not sure exactly what you mean by a "proper workaround"?

With regard to porting this fix to 2.1, the bar is very high and right now this issue does not meet that bar.

**Comment by spottedmahn** [comment]
>The workaround you show above is what we would recommend as a workaround

Copy that.  Thanks for your response! 😊

>With regard to porting this fix to 2.1, the bar is very high and right now this issue does not meet that bar.

Gotcha.  Wonder what is the LTS bar?  Wonder if that is documented anywhere? 🤔
